docs/eBpfExtensions.md

```diff
@@ -1,42 +1,207 @@
-[Network Module Registrar (NMR)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/network-module-registrar2)
-to decouple the eBPF extensions from the core eBPF for Windows framework.  In the NMR architecture,
+[Network Module Registrar (NMR)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/introduction-to-the-network-module-registrar)
+feature in Windows to develop eBPF extensions that are decoupled from the core eBPF for Windows framework.  NMR facilitates software
+modules to be attached to each other. The
 [Network Programming Interface (NPI)](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/network-programming-interface)
-contracts are identified by a GUID.
+defines the contract between the modules.  The eBPF extensions would implement the
+[Provider Modules](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/provider-module) for two types of NPIs as discussed in detail below.
+Even though the NMR component and the relevant APIs have the word "Network" in them, these software modules
+are completely agnostic of Windows Networking stack and non-networking eBPF extensions can be developed using these.
 
-## NPI Contract for eBPF Program Information
+## 1.2 Prerequisites
+The audience is assumed to be familiar with [NMR Architecture](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/architecture-overview) and
+the various aspects of developing NMR modules as documented [here](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/network-module-registrar2).
 
+## 1.3 NPI Contracts for eBPF Extensions
+eBPF Extensions need to implement *provider modules* for two types of NPIs. They are the **Program Information NPI** provider and the **Hook NPI** provider.
+
+### 1.3.1 eBPF Program Information NPI Provider
 The eBPF program information NPI contract is used to provide information about an eBPF program type. Program types
 are defined as the ABI contracts that eBPF programs are written to conform to.
 This information is consumed by the eBPF verifier to ensure that any eBPF programs of a given type are safe to load
 and execute. The ABI contract includes both a description of the &quot;context&quot; parameter passed to the eBPF
-program as well as the list of specific helper functions that are available to such eBPF programs. eBPF extensions
-should not unload until the last NMR client detaches, signifying that no eBPF programs are using helper functions
-it provides.
+program as well as the list of specific helper functions that are available to such eBPF programs.
+
+eBPF extensions must implement a different Program Information provider module for each program type for which it provides
+implementation of hooks or helper functions specific to that program type.
+
+### 1.3.2 eBPF Hook NPI Provider
+The Hook NPI contracts are used by extension to invoke eBPF programs when OS events occur. These are also known as *Attach NPI* and is used
```

Comment:
why do we need two terms?  Can we just use one throughout given the NPI terminology is unique to Windows?

Reply:
Sticking to "Hook NPI provider" and "attach types".